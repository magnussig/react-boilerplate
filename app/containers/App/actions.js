/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS,
  LOAD_REPOS_ERROR,
  LOAD_USERS_SUCCESS,
  LOAD_USERS,
  LOAD_USERS_ERROR,
  SET_ALL_TEAMS,
  SET_ALL_USERS,
  SET_CURR_TEAM_LEAD,
  CHANGE_USERNAME,
} from './constants';

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_REPOS
 */
export function loadRepos(teamId) {
  return {
    type: LOAD_REPOS,
    teamId,
  };
}

/**
 * Dispatched when the repositories are loaded by the request saga
 *
 * @param  {array} repos The repository data
 * @param  {string} username The current username
 *
 * @return {object}      An action object with a type of LOAD_REPOS_SUCCESS passing the repos
 */
export function reposLoaded(repos, username) {
  return {
    type: LOAD_REPOS_SUCCESS,
    repos,
    username,
  };
}

/**
 * Dispatched when loading the repositories fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_REPOS_ERROR passing the error
 */
export function repoLoadingError(error) {
  return {
    type: LOAD_REPOS_ERROR,
    error,
  };
}

export function loadUsers(userId) {
  return {
    type: LOAD_USERS,
    userId,
  };
}

export function userLoaded(users) {
  return {
    type: LOAD_USERS_SUCCESS,
    users,
  };
}

export function userLoadingError(error) {
  return {
    type: LOAD_USERS_ERROR,
    error,
  };
}

export function setAllTeams(teams) {
  return {
    type: SET_ALL_TEAMS,
    teams,
  };
}

export function setAllUsers(users) {
  return {
    type: SET_ALL_USERS,
    users,
  };
}

export function setCurrTeamLead(lead) {
  return {
    type: SET_CURR_TEAM_LEAD,
    lead,
  };
}

export function changeUsername(username) {
  return {
    type: CHANGE_USERNAME,
    username,
  };
}
