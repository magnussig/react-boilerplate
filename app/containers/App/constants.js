/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_USERS_SUCCESS = 'boilerplate/App/LOAD_USERS_SUCCESS';
export const LOAD_USERS_ERROR = 'boilerplate/App/LOAD_USERS_ERROR';
export const LOAD_USERS = 'boilerplate/App/LOAD_USERS';
export const LOAD_INFO = 'boilerplate/App/LOAD_INFO';
export const LOAD_INFO_SUCCESS = 'boilerplate/App/LOAD_INFO_SUCCESS';
export const LOAD_INFO_ERROR = 'boilerplate/App/LOAD_INFO_ERROR';
export const SET_ALL_TEAMS = 'boilerplate/App/SET_ALL_TEAMS';
export const SET_ALL_USERS = 'boilerplate/App/SET_ALL_USERS';
export const SET_CURR_TEAM_LEAD = 'boilerplate/App/SET_CURR_TEAM_LEAD';
export const CHANGE_USERNAME = 'boilerplate/App/CHANGE_USERNAME';
