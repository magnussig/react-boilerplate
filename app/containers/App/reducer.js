/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_REPOS_SUCCESS,
  LOAD_REPOS,
  LOAD_REPOS_ERROR,
  LOAD_USERS_SUCCESS,
  LOAD_USERS,
  LOAD_USERS_ERROR,
  SET_ALL_TEAMS,
  SET_ALL_USERS,
  SET_CURR_TEAM_LEAD,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentUser: false,
  userData: {
    repositories: false,
  },
  teamId: null,
  users: null,
  userId: null,
  allTeams: null,
  allUsers: null,
  currTeamLead: null,
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_REPOS:
      return state
        .set('loading', true)
        .set('error', false)
        .set('teamId', action.teamId)
        .setIn(['userData', 'repositories'], false);
    case LOAD_REPOS_SUCCESS:
      return state
        .setIn(['userData', 'repositories'], action.repos)
        .set('loading', false)
        .set('currentUser', action.username);
    case LOAD_REPOS_ERROR:
      return state.set('error', action.error).set('loading', false);
    case LOAD_USERS:
      return state
        .set('loading', true)
        .set('error', false)
        .set('userId', action.userId)
        .set('users', null);
    case LOAD_USERS_SUCCESS:
      return state.set('users', action.users).set('loading', false);
    case LOAD_USERS_ERROR:
      return state.set('error', action.error).set('loading', false);
    case SET_ALL_TEAMS:
      return state.set('allTeams', action.teams);
    case SET_ALL_USERS:
      return state.set('allUsers', action.users);
    case SET_CURR_TEAM_LEAD:
      return state.set('currTeamLead', action.lead);
    default:
      return state;
  }
}

export default appReducer;
