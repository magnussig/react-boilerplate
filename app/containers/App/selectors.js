/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = state => state.get('global');

const selectRouter = state => state.get('router');

const makeSelectCurrentUser = () =>
  createSelector(selectGlobal, globalState => globalState.get('currentUser'));

const makeSelectLoading = () =>
  createSelector(selectGlobal, globalState => globalState.get('loading'));

const makeSelectError = () =>
  createSelector(selectGlobal, globalState => globalState.get('error'));

const makeSelectRepos = () =>
  createSelector(selectGlobal, globalState =>
    globalState.getIn(['userData', 'repositories']),
  );

const makeSelectLocation = () =>
  createSelector(selectRouter, routerState =>
    routerState.get('location').toJS(),
  );

const makeSelectTeam = () =>
  createSelector(selectGlobal, globalState => globalState.get('teamId'));

const makeSelectUsers = () =>
  createSelector(selectGlobal, globalState => globalState.get('users'));

const makeSelectUserId = () =>
  createSelector(selectGlobal, globalState => globalState.get('userId'));

const makeSelectAllTeams = () =>
  createSelector(selectGlobal, globalState => globalState.get('allTeams'));

const makeSelectAllUsers = () =>
  createSelector(selectGlobal, globalState => globalState.get('allUsers'));

const makeSelectCurrTeamLead = () =>
  createSelector(selectGlobal, globalState => globalState.get('currTeamLead'));

export {
  selectGlobal,
  makeSelectCurrentUser,
  makeSelectLoading,
  makeSelectError,
  makeSelectRepos,
  makeSelectLocation,
  makeSelectTeam,
  makeSelectUsers,
  makeSelectUserId,
  makeSelectAllTeams,
  makeSelectAllUsers,
  makeSelectCurrTeamLead,
};
