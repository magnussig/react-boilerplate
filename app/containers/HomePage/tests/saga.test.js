/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_INFO } from 'containers/App/constants';
import { infoLoaded, infoLoadingError } from 'containers/App/actions';

import githubData, { getInfo } from '../saga';

const username = '';

/* eslint-disable redux-saga/yield-effects */
describe('getInfo Saga', () => {
  let getInfoGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getInfoGenerator = getInfo();

    const selectDescriptor = getInfoGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getInfoGenerator.next(username).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  /*it('should dispatch the infoLoaded action if it requests the data successfully', () => {
    const response = [
      {
        name: 'first team',
      },
      {
        name: 'second team',
      },
    ];
    const putDescriptor = getInfoGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(infoLoaded(response, username)));
  });*/

  it('should call the infoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getInfoGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(infoLoadingError(response)));
  });
});

describe('githubDataSaga Saga', () => {
  const githubDataSaga = githubData();

  it('should start task to watch for LOAD_INFO action', () => {
    const takeLatestDescriptor = githubDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_INFO, getInfo));
  });
});
