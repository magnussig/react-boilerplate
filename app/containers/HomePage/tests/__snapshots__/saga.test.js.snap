// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInfo Saga should call the infoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getInfo Saga should call the infoLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://tempo-test.herokuapp.com/7d1d085e-dbee-4483-aa29-ca033ccae1e4/1/team/",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getInfo Saga should dispatch the infoLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getInfo Saga should dispatch the infoLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://tempo-test.herokuapp.com/7d1d085e-dbee-4483-aa29-ca033ccae1e4/1/team/",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
